argspec - fix validating type for required options that are None

s-hertel opened the pull request:
##### SUMMARY
Fixes #79656

Fix skipping validating the type for all options that are None, even when None is not the default (e.g. when the option is required).

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
ArgumentSpecValidator

s-hertel committed:
Only bypass type validation for null parameters if the default is None. A default is mutually exclusive with required.

s-hertel committed:
changelog

bcoca commented on lib/ansible/module_utils/common/parameters.py:
```diff
@@ -600,7 +600,7 @@ def _validate_argument_types(argument_spec, parameters, prefix='', options_conte
 
         value = parameters[param]
-        if value is None:
+        if value is None and not spec.get('required') and spec.get('default') is None:
```
2 other things to consider `choices` and type `raw`

s-hertel replied:
I have a test for raw validating that None is a valid value for a required raw option.

`choices` are checked after type coercion, and we don't validate `choices` match the `type`, so I think this improves internally consistency. This spec `dict(required=True, choices=[None, 'other_choice'])` would previously allow None. But it's ambiguous because the type is 'str', so imo it should be either
`dict(required=True, type='raw', choices=[None, 'other_choice'])`
or
`dict(required=True, choices=['None', 'other_choice'])`

s-hertel committed:
replace question with link to overlapping feature request

s-hertel committed:
ci_complete

s-hertel committed:
don't convert None to 'None' for str options, as it's not backwards compatible

simplify required str options so modules don't need to check if a required str option is None

This is possibly a breaking change for 3rd party modules, but for modules that do not guard against None explicitly provided to required options it would be an improvement.
Alternatively, check_type_str could succeed for None and return it as-is, and that would be backwards compatible but reintroduce https://github.com/ansible/ansible/issues/79656 for string types.

s-hertel committed:
put check back for empty str

s-hertel commented on lib/ansible/module_utils/common/validation.py:
```diff
@@ -381,7 +381,7 @@ def check_type_str(value, allow_conversion=True, param=None, prefix=''):
         return value
 
-    if allow_conversion:
+    if allow_conversion and value is not None:
```
Previous to the change in lib/ansible/module_utils/common/parameters.py, `None` string options were never converted to `'None'` because the type enforcement was skipped for all None values. However with my change, the type is now validated for required options and options that have a non-None default.

Returning 'None' would be a breaking change for `apt_repository`, which relies on None as a valid value for the **required** str option [`repo`](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/apt_repository_module.html#parameter-repo).

I waffled between two ways to fix this. Either do this, and update the `apt_repository` tests since `None` will now fail type conversion (what I did), or return None as-is just above (requires no update to the tests, but I wasn't sure if it makes sense for `check_type_str` to return a Nonetype).

```python
    if isinstance(value, string_types) or value is None:
        return value
```

felixfontein commented:
This PR only seems to cover `required`, but not `required_together`, `required_if`, etc., reported in #69190, which have the same problem. (See also my PR #72248 from 2020.)

s-hertel commented:
@felixfontein They're related issues, but I don't think this really replaces anything in #72248. This enforces (non-str) type-validation for required options (it was always skipped for None), but it does not directly disallow None for required options or attempt to fix the more general issue in https://github.com/ansible/ansible/issues/69190.

felixfontein commented:
@s-hertel looks like I didn't look at this closely enough. Sorry for the noise :)

s-hertel commented:
rebuild_merge
